pysipp/launch.py

```diff
@@ -21,116 +22,109 @@ class TimeoutError(Exception):
-        self.spm = subprocmod
-        self.osm = osmod
-        self.poller = poller()
-        # collector thread placeholder
-        self._waiter = None
         # store proc results
         self._procs = OrderedDict()
 
-    def __call__(self, cmds, block=True, rate=300, **kwargs):
-        if self._waiter and self._waiter.is_alive():
+    async def run(
+        self,
+        cmds,
+        rate=300,
+        **kwargs
+    ):
+        if self.is_alive():
             raise RuntimeError(
                 "Not all processes from a prior run have completed"
             )
         if self._procs:
             raise RuntimeError(
                 "Process results have not been cleared from previous run"
             )
-        sp = self.spm
-        os = self.osm
-        DEVNULL = open(os.devnull, 'wb')
-        fds2procs = OrderedDict()
-
         # run agent commands in sequence
         for cmd in cmds:
             log.debug(
                 "launching cmd:\n\"{}\"\n".format(cmd))
-            proc = sp.Popen(
+            # proc = await trio.open_process(
+            proc = trio.Process(
                 shlex.split(cmd),
-                stdout=DEVNULL,
-                stderr=sp.PIPE
+                stdout=subprocess.DEVNULL,
+                stderr=subprocess.PIPE
             )
-            fd = proc.stderr.fileno()
-            log.debug("registering fd '{}' for pid '{}'".format(
-                fd, proc.pid))
-            fds2procs[fd] = self._procs[cmd] = proc
-            # register for stderr hangup events
-            self.poller.register(proc.stderr.fileno(), select.EPOLLHUP)
+            self._procs[cmd] = proc
+
             # limit launch rate
             time.sleep(1. / rate)
 
-        # launch waiter
-        self._waiter = threading.Thread(target=self._wait, args=(fds2procs,))
-        self._waiter.daemon = True
-        self._waiter.start()
-
-        return self.get(**kwargs) if block else self._procs
+        return self._procs
 
-    def _wait(self, fds2procs):
-        log.debug("started waiter for procs {}".format(fds2procs))
+    async def get(self, timeout=180):
+        '''Block up to `timeout` seconds for all agents to complete.
+        Either return (cmd, proc) pairs or raise `TimeoutError` on timeout
+        '''
         signalled = None
-        left = len(fds2procs)
-        collected = 0
-        while collected < left:
-            pairs = self.poller.poll()  # wait on hangup events
-            log.debug("received hangup for pairs '{}'".format(pairs))
-            for fd, status in pairs:
-                collected += 1
-                proc = fds2procs[fd]
-                # attach streams so they can be read more then once
-                log.debug("collecting streams for {}".format(proc))
-                proc.streams = Streams(*proc.communicate())  # timeout=2))
-                if proc.returncode != 0 and not signalled:
+
+        # taken mostly verbatim from ``trio.run_process()``
```

Comment:
There's also new work in python-trio/trio#1568 that will likely make some of this simpler.